xarray/tests/test_backends_datatree.py

```diff
@@ -539,6 +539,23 @@ def test_phony_dims_warning(self, tmpdir) -> None:
                     "phony_dim_3": 25,
                 }
 
+    def test_roundtrip_via_bytes(self, simple_datatree):
+        original_dt = simple_datatree
+        roundtrip_dt = open_datatree(original_dt.to_netcdf())
+        assert_equal(original_dt, roundtrip_dt)
```

Comment:
Could you add another test to veirfy that `dataset.to_netcdf(engine='h5netcdf')` works?

Reply:
Yes this works too, done.

Reply:
I was suggesting adding this test for `Dataset.to_netcdf()` rather than `DataTree.to_netcdf()`, but I'm glad that also works!

Reply:
OK I've added this test too -- good spot as it flagged another issue hiding in the existing code that worked fine by lucky coincidence with scipy but not h5netcdf:

`Dataset.to_netcdf(engine='h5netcdf')` was getting the bytes before the h5py file was closed, meaning the magic number didn't get written and so the bytes were invalid. (Maybe this is why the existing code seemed to think that h5netcdf couldn't handle returning bytes?)

The fix for that was to call BytesIO.getvalue after the store has closed, but that then broke the scipy backend, which closes the BytesIO on close of its store, preventing BytesIO.getvalue from being called. I have worked around that by preventing BytesIO.close from being called.

Reply:
Resolving as this is now fixed via a combination of my and @shoyer's commits.